### PR TITLE
Use go modules, update controller-{runtime, tools} to v2

### LINKED ISSUE #1025: update controller-runtime
/kind feature

This is a tracking issue for people to pay attention to the controller-runtime dependency update.

**Anything else you would like to add:**
Add related issues below and why they are related:

* https://github.com/kubernetes-sigs/cluster-api/issues/296 -- crd web hook validations are coming with the next version of controller runtime.

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
- Upgrades `controller-runtime` to `v0.2.0-beta.2`.
- Upgrades `controller-tools` to `v0.2.0-beta.2`.
- Upgrades CI fetchtools Kubebuilder to 1.14.1 release.
- Removes `dep`, `Gopkg.lock`, and `Gopkg.toml`.
- Overhauls Makefile to use new `controller-tools` utilities.
- Adds `hack/tools.go` file to force-vendor utilities.
- Adds `hack/ensure-go.sh` to check and setup Go for modules.
- Removes `-` from controller names (e.g. `machine-controller` -> `machine_controller`).
- Updates integration tests to use `kubectl@1.14.1` and `kind@0.3.0`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1025 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: hack/update-vendor.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+export GOPATH=
+# Explicitly clear GOFLAGS, since GOFLAGS=-mod=vendor breaks dependency resolution while rebuilding vendor
+export GOFLAGS=
+# Detect problematic GOPROXY settings that prevent lookup of dependencies
+if [[ "${GOPROXY:-}" == "off" ]]; then
+  echo "Cannot run hack/update-vendor.sh with \$GOPROXY=off"
+  exit 1
+fi
+
+go mod vendor -v
```
Comment (akutz) [inline hack/update-vendor.sh]: Hi @vincepri,

Please consider changing this to:

```shell
go mod tidy -v
go mod vendor -v
go mod verify
```

**Comment by vincepri** [comment]
Current blockers:
- [ ] CI images have an outdated version of Go (<1.12.1) [PR](https://github.com/kubernetes/test-infra/pull/13113).
- [ ] Failing tests on Update.Status (details [here](https://kubernetes.slack.com/archives/CAR30FCJZ/p1561092115400800))

Non blocking issues:
- [x] RBAC tags have to be placed at the top of the file, without any code around it to work.
- [ ] RBAC RoleBinding isn't generated from KubeBuilder/controller-tools anymore.

File: hack/update-vendor.sh (hunk 1 above)
Comment (vincepri) [reply hack/update-vendor.sh]: Done!

**Comment by vincepri** [comment]
/hold

Currently waiting for directions on how to handle the ObjectMeta embedding issue from @DirectXMan12 and others. We can decide to merge this as it is with the fork to temporarily unblock v1alpha2 progress or wait if the timeframe to fix it in controller-tools works.

File: hack/pin-dependency.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  echo ""
+  exit 1
+fi
+
+# Add the require directive
+echo "Running: go get ${dep}@${sha}"
+go get "${dep}@${sha}"
+
+# Find the resolved version
+rev=$(go mod edit -json | jq -r '.Require[] | select(.Path == "github.com/docker/docker") | .Version')
```
Comment (ncdc) [inline hack/pin-dependency.sh]: I'm pretty sure we don't want to hardcode docker in here?

**Comment by ncdc** [comment]
I have at least 1 pending review comment that will need fixing before we merge, but I'm still making my way through the diffs. Until I get that submitted:

/hold

File: pkg/controller/cluster/cluster_controller.go [hunk 3]
```diff
@@ -50,7 +50,7 @@ func newReconciler(mgr manager.Manager, actuator Actuator) reconcile.Reconciler
 // add adds a new Controller to mgr with r as the reconcile.Reconciler
 func add(mgr manager.Manager, r reconcile.Reconciler) error {
 	// Create a new controller
-	c, err := controller.New("cluster-controller", mgr, controller.Options{Reconciler: r})
+	c, err := controller.New("cluster_controller", mgr, controller.Options{Reconciler: r})
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: Is `-` disallowed by the new controller-runtime?

Comment (vincepri) [reply pkg/controller/cluster/cluster_controller.go]: Not by controller-runtime, but opencensus uses this name to make the metrics and in the newer version dashes aren't allowed

It also matches the go file, so win-win? :D

File: hack/pin-dependency.sh (hunk 2 above)
Comment (vincepri) [reply hack/pin-dependency.sh]: True, thanks for catching that!

File: cmd/clusterctl/validation/validate_cluster_api_objects.go [hunk 4]
```diff
@@ -43,7 +43,7 @@ func ValidateClusterAPIObjects(ctx context.Context, w io.Writer, c client.Client
 	}
 
 	machines := &clusterv1alpha1.MachineList{}
-	if err := c.List(ctx, client.InNamespace(namespace), machines); err != nil {
+	if err := c.List(ctx, machines, client.InNamespace(namespace)); err != nil {
```
Comment (detiber) [inline cmd/clusterctl/validation/validate_cluster_api_objects.go]: What caused the ordering to change here? Is it just related to the updated client generation?

File: pkg/client/clientset_generated/clientset/clientset.go [hunk 5]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Code generated by client-gen. DO NOT EDIT.
+// Code generated by main. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/client/clientset_generated/clientset/clientset.go]: Yes

File: cmd/clusterctl/validation/validate_cluster_api_objects.go (hunk 4 above)
Comment (ncdc) [reply cmd/clusterctl/validation/validate_cluster_api_objects.go]: It's from a controller-runtime change to the method signature

Comment (ncdc) [reply cmd/clusterctl/validation/validate_cluster_api_objects.go]: https://github.com/kubernetes-sigs/controller-runtime/pull/294

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -20,11 +20,18 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.k8s.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.k8s.io,resources=machinedeployments;machinedeployments/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.k8s.io,resources=machinesets;machinesets/status,verbs=get;list;watch;create;update;patch;delete
```
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/controller/controller.go]: They don't work if defined elsewhere, that's why I moved them in here

File: cmd/clusterctl/testdata/create-cluster-no-args-invalid-flag.golden [hunk 7]
```diff
@@ -25,7 +25,7 @@ Global Flags:
       --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
       --log-flush-frequency duration     Maximum number of seconds between log flushes (default 5s)
       --logtostderr                      log to standard error instead of files (default true)
-      --master string                    The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.
+      --master --kubeconfig              (Deprecated: switch to --kubeconfig) The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.
```
[the remark this answers is not in the document]
Comment (vincepri) [reply cmd/clusterctl/testdata/create-cluster-no-args-invalid-flag.golden]: I checked with a master build of `clusterctl` and it seems that's just how the help is printed out

File: docs/book/provider_implementations/overview.md [hunk 8]
```diff
@@ -1,14 +1,13 @@
 # Overview
 
-In order to demonstrate how to develop a new Cluster API provider we will use 
+In order to demonstrate how to develop a new Cluster API provider we will use
 `kubebuilder` to create an example provider. For more information on `kubebuilder`
 and CRDs in general we highly recommend reading the [Kubebuilder Book][kubebuilder-book].
 Much of the information here was adapted directly from it. The minimal version of
-`kubebuilder` required is [`v1.0.5`][kubebuilder-1.0.5].
+`kubebuilder` required is [`v2.0.0`][kubebuilder-2.0.0].
```
[the remark this answers is not in the document]
Comment (vincepri) [reply docs/book/provider_implementations/overview.md]: Opened https://github.com/kubernetes-sigs/cluster-api/issues/1075

File: cmd/clusterctl/testdata/create-cluster-no-args-invalid-flag.golden (hunk 7 above)
[the remark this answers is not in the document]
Comment (detiber) [reply cmd/clusterctl/testdata/create-cluster-no-args-invalid-flag.golden]: right, this seems broken

File: pkg/controller/controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (DirectXMan12) [reply pkg/controller/controller.go]: they should work anywhere in the packages that controller-gen is told to read, as long at they're not in, say, the Godoc for a function or something.

File: cmd/clusterctl/validation/validate_cluster_api_objects.go (hunk 4 above)
Comment (DirectXMan12) [reply cmd/clusterctl/validation/validate_cluster_api_objects.go]: we changed the method signature to be in line with other methods and to support easier option passing.

**Pull request merged**